---
title: "Generic annotation"
author: "Produced with bacannot pipeline"
date: "`r format(Sys.time(), '%d %B %Y')`"
params:
  generic_annotation:
  generic_annotator:
  kegg:
  barrnap:
  mlst:
  refseq_masher:
  query:
output:
  bookdown::html_document2:
    css: styles.css
    df_print: paged
    number_sections: false
    syntax: espresso
    self_contained: yes
    theme: readable
    code_folding: "show"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, fig.pos = 'H')
suppressMessages(library(magrittr))
suppressMessages(library(knitr))
suppressMessages(library(DT))
suppressMessages(library(ballgown))
suppressMessages(library(stringr))
suppressMessages(library(dplyr))

# Line checker
check_lines <- function(x) {
  return(ifelse(identical(as.integer(nrow(x)), integer(0)), 0, nrow(x)))
}

# Load data
if (params$generic_annotator == "bakta") {
  generic_annotation_txt <- try(read.delim(params$generic_annotation, header = FALSE, sep = ":"), silent = TRUE) %>% filter(!V1=="Sequence(s)") %>% filter(!V1=="Annotation") %>% filter(!V1=="Bakta")
} else {
  generic_annotation_txt <- try(read.delim(params$generic_annotation, header = FALSE, sep = ":"), silent = TRUE)
}
mlst_txt <- try(read.delim(params$mlst, header = FALSE), silent = TRUE)
barrnap_gff <- try(gffRead(params$barrnap), silent = TRUE)
if (class(barrnap_gff) != 'try-error' &
    check_lines(barrnap_gff) > 0) {
      barrnap_gff <- barrnap_gff
} else {
  barrnap_gff <- data.frame()
}

refseq_masher_txt <- try(read.delim(params$refseq_masher, header = TRUE), silent = TRUE)
if (file.exists(params$kegg)) {
  kegg_not_null <- TRUE
  kegg_svg <- params$kegg
} else {
  kegg_not_null <- FALSE
}

if (params$generic_annotator == "prokka") {
  annotator_url <- "https://github.com/tseemann/prokka"
  prokka_not_null <- TRUE
}
if (params$generic_annotator == "bakta") {
  annotator_url <- "https://github.com/oschwengers/bakta"
  prokka_not_null <- FALSE
}

# DT options
# Lists
dt_opt_lst <- list(pageLength = 1,
                         lengthMenu = c(1, 5, 10),
                         dom='flrtBip',
                         buttons = c('copy', 'csv', 'excel', 'colvis'),
                         scrollX = TRUE,
                         fixedColumns = FALSE,
                         autoWidth = TRUE,
                         columnDefs = list(list(className = 'dt-body-left', targets = -1)))

# tabset
# {.tabset .tabset-pills}
```

# `r params$query` general annotation report

## About

This report was built to summarise in a report the results of the most generic annotation contents, which are: `r str_to_title(params$generic_annotator)`, Barrnap, mlst, KofamScan and refseq_masher. If you'd like to see any other result included in this report please flag an [enhancement issue on Github](https://github.com/fmalmeida/bacannot/issues).

## RefSeq Masher

[RefSeq Masher](https://github.com/phac-nml/refseq_masher) is a tool that enables to rapidly find what NCBI RefSeq genomes match or are contained within your sequence data using Mash MinHash with a Mash sketch database of NCBI RefSeq Genomes. The results are shown below (bacannot outputs only the top 10).

```{r refseq_masher, echo=FALSE}
datatable(refseq_masher_txt,
          escape = FALSE,
          filter = 'top',
          options = dt_opt_lst,
          extensions = 'Buttons',
          rownames = F
        )
```


## MLST

Bacannot uses the [mlst](https://github.com/tseemann/mlst) package to scan the PubMLST schemes available in order to classify the genome under public multilocus sequence type schemes. The results for `r params$query` are shown below.

```{r mlst_stats, echo=FALSE}
sketch = htmltools::withTags(table(
  class = 'display',
  thead(
    tr(
      th(rowspan = 1, 'Assembly'),
      th(rowspan = 1, 'Scheme'),
      th(rowspan = 1, 'ST'),
      th(colspan = ncol(mlst_txt)-3, 'Alleles')
    )
  )
))
datatable(mlst_txt,
          escape = FALSE,
          filter = 'top',
          container = sketch,
          options = dt_opt_lst,
          extensions = 'Buttons',
          rownames = F
        )
```

## `r str_to_title(params$generic_annotator)`

[`r str_to_title(params$generic_annotator)`](`r annotator_url`) is generic prokaryotic genome annotation tool that produces standards-compliant output files.

```{r kegg_svg, echo=FALSE, results='asis', eval=prokka_not_null}
> In bacannot, the prokka database is incremented with TIGRFAM's hmm [hosted at NCBI](https://ftp.ncbi.nlm.nih.gov/hmm/TIGRFAMs/release_15.0/).
```

```{r generic_stats, echo=FALSE}
df <- data.frame(generic_annotation_txt[,-1])
rownames(df) <- generic_annotation_txt[,1]
datatable(t(df),
          escape = FALSE,
          filter = 'top',
          options = dt_opt_lst,
          extensions = 'Buttons',
          rownames = F
        )
```

## Barrnap

[Barrnap](https://github.com/tseemann/barrnap) is a fast Ribosomal RNA predictor for bacterias, from the same developer of Prokka. It is fast and produces a GFF of the predicted rRNAs (See below).

```{r barrnap_gff, echo=FALSE}
datatable(barrnap_gff,
          escape = FALSE,
          filter = 'top',
          options = dt_opt_lst,
          extensions = 'Buttons',
          rownames = F
          )
```


```{r kegg_svg, echo=FALSE, results='asis', eval=kegg_not_null, child='yes_kegg.Rmd'}
```
