---
title: "Annotation of mobile genetic elements"
author: "Produced with bacannot pipeline"
date: "`r format(Sys.time(), '%d %B %Y')`"
params:
  blast_id:
  blast_cov:
  custom_blast:
  blast_db:
  blast_gff:
  query:
output:
  bookdown::html_document2: 
    css: styles.css
    df_print: paged
    number_sections: false
    syntax: espresso
    self_contained: yes
    theme: readable
    code_folding: "show"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, fig.pos = 'H')
suppressMessages(library(magrittr))
suppressMessages(library(knitr))
suppressMessages(library(tidyr))
suppressMessages(library(plyr))
suppressMessages(library(DT))
suppressMessages(library(dplyr))
suppressMessages(library(stringr))
suppressMessages(library(ggplot2))

# Line checker
check_lines <- function(x) {
  return(ifelse(identical(as.integer(nrow(x)), integer(0)), 0, nrow(x)))
}

## Read ICEberg documents
custom_blast <- try(read.delim(params$custom_blast, header = TRUE), silent = TRUE)
blast_gff <- try(read.delim(params$blast_gff, header = FALSE), silent = TRUE)

## Check for emptyness
if (class(custom_blast) == "try-error" ||
    check_lines(custom_blast) == 0) {
  custom_blast <- data.frame(
    matrix(ncol = 14, nrow = 0)
  )  
} else {
  custom_blast[sapply(custom_blast, is.character)] <- 
    lapply(custom_blast[sapply(custom_blast, is.character)], as.factor)
}

if (class(blast_gff) == "try-error" ||
    check_lines(blast_gff) == 0) {
  blast_gff <- data.frame(
    matrix(ncol = 9, nrow = 0)
  )  
} else {
  blast_gff[sapply(blast_gff, is.character)] <- 
    lapply(blast_gff[sapply(blast_gff, is.character)], as.factor)
}

# DT options
# Lists
dt_opt_lst <- list(pageLength = 5,
                         lengthMenu = c(5, 10, 15, 20, 50),
                         dom='flrtBip',
                         buttons = c('copy', 'csv', 'excel', 'colvis'),
                         scrollX = TRUE,
                         fixedColumns = FALSE,
                         autoWidth = TRUE)
```

# `r params$query` report of user's custom database annotation: `r params$blast_db`

## About

This report is based on the user's custom database input called `r params$blast_db`. This database have been blasted against the query genome via BLASTn. The BLASTn raw results are shown in the table \@ref(tab:blastn-raw-table). Additionally, the BLAST results have been used to detect intersection points with the annotation using the software bedtools. Therefore, in table \@ref(tab:blastn-gff-table), we show the gene features from the query genome that have intersection points with the BLAST results.

> Take note that DT tables will only be rendered when input is available. Thus, whenever a table is missing it is because no results have been found.

### BLAST prediction thresholds

All the predictions were passed through a user defined threshold for minimum coverage and identity:

* Min. Identity (%): `> `r params$blast_id``
* Min. Coverage (%): `> `r params$blast_cov``

## Results

### BLASTn results

The BLASTn results of the custom database against the query genome is shown in the table \@ref(tab:blastn-raw-table).

<br>
<caption>(#tab:blastn-raw-table) Alignment of the `r params$blast_db` custom database against the query genome via BLASTn</caption>
```{r}
blast <- custom_blast
# Render dt
datatable(blast,
          escape = FALSE,
          filter = 'top',
          options = dt_opt_lst,
          extensions = 'Buttons',
          rownames = F)
```

### Annotation intersection

Using the software bedtools intersect, the BLASTn results have been used to search for intersections against the query gene annotation. This result is shown in table \@ref(tab:blastn-gff-table).

<br>
<caption>(#tab:blastn-gff-table) Annotation intersection of the `r params$blast_db` custom database BLASTn results</caption>
```{r}
blast <- blast_gff
colnames(blast) <-
  c("Contig", "Source", "Feature", "Start", "End", "Score", "Strand", "Phase", "Attributes")
# Render dt
datatable(blast,
          escape = FALSE,
          filter = 'top',
          options = dt_opt_lst,
          extensions = 'Buttons',
          rownames = F)
```
